ecs-cli/modules/clients/aws/servicediscovery/client_test.go

```diff
@@ -0,0 +1,274 @@
+					VPCRegion: aws.String("ap-south-1"),
+				},
+			},
+		},
+		"zone4": route53.GetHostedZoneOutput{
+			VPCs: []*route53.VPC{
+				&route53.VPC{
+					VPCId:     aws.String("vpc-C00010FF"),
+					VPCRegion: aws.String("sa-east-1"),
+				},
+			},
+		},
+		"zone5": route53.GetHostedZoneOutput{
+			VPCs: []*route53.VPC{
+				&route53.VPC{
+					VPCId:     aws.String("vpc-DEADBAAD"),
+					VPCRegion: aws.String("ap-south-1"),
+				},
+				&route53.VPC{
+					VPCId:     aws.String("vpc-D15EA5E"),
+					VPCRegion: aws.String("ap-south-1"),
+				},
+			},
+		},
+	}
+
+	// Mock API actions
+	getHostedZone = func(id *string, params *config.CommandConfig) (*route53.GetHostedZoneOutput, error) {
+		zone := hostedZoneData[aws.StringValue(id)]
+		return &zone, nil
+	}
+}
+
+func setupNamespaceMocks(t *testing.T, publicNamespaces bool) {
+	// Mock Data
+	namespaceType := servicediscovery.NamespaceTypeDnsPrivate
+	if publicNamespaces {
+		namespaceType = servicediscovery.NamespaceTypeDnsPublic
+	}
+	var namespaceData = map[string]servicediscovery.Namespace{
+		"namespace1": servicediscovery.Namespace{
+			Id:   aws.String("namespace1"),
+			Name: aws.String("corp"),
+			Properties: &servicediscovery.NamespaceProperties{
+				DnsProperties: &servicediscovery.DnsProperties{
+					HostedZoneId: aws.String("zone1"),
+				},
+			},
+			Type: aws.String(namespaceType),
+		},
+		"namespace2": servicediscovery.Namespace{
+			Id:   aws.String("namespace2"),
+			Name: aws.String("prod"),
+			Properties: &servicediscovery.NamespaceProperties{
+				DnsProperties: &servicediscovery.DnsProperties{
+					HostedZoneId: aws.String("zone2"),
+				},
+			},
+			Type: aws.String(namespaceType),
+		},
+		"namespace3": servicediscovery.Namespace{
+			Id:   aws.String("namespace3"),
+			Name: aws.String("bridge"),
+			Properties: &servicediscovery.NamespaceProperties{
+				DnsProperties: &servicediscovery.DnsProperties{
+					HostedZoneId: aws.String("zone3"),
+				},
+			},
+			Type: aws.String(namespaceType),
+		},
+	}
+
+	if !publicNamespaces {
+		// Add in extra namespaces with the same name, since this is possible with private namespaces
+		namespaceData["namespace4"] = servicediscovery.Namespace{
+			Id:   aws.String("namespace4"),
+			Name: aws.String("corp"),
+			Properties: &servicediscovery.NamespaceProperties{
+				DnsProperties: &servicediscovery.DnsProperties{
+					HostedZoneId: aws.String("zone4"),
+				},
+			},
+			Type: aws.String(namespaceType),
+		}
+		namespaceData["namespace5"] = servicediscovery.Namespace{
+			Id:   aws.String("namespace5"),
+			Name: aws.String("bridge"),
+			Properties: &servicediscovery.NamespaceProperties{
+				DnsProperties: &servicediscovery.DnsProperties{
+					HostedZoneId: aws.String("zone5"),
+				},
+			},
+			Type: aws.String(namespaceType),
+		}
+
+	}
+
+	// Mock API Actions
+	listNamespaces = func(isPublic bool, params *config.CommandConfig, fn func(*servicediscovery.NamespaceSummary) bool) error {
```

Comment:
I'm not sure I'm a fan of  this pattern of mocking out this private function here. Since the thing we're trying to mock is the client, it seems like in the implementation we could perhaps use the pattern of setting a function to a variable on the client constructor/factory/getter, so that that is easier to mock. 

Mocking the function this way also seems to couple the test very tightly with the implementation, and so we aren't actually testing that the function is doing what we expect.

Reply:
I am not trying to test the listNamespaces function, because that is a thin wrapper of over ListNamespacePages in the SDK as mentioned elsewhere. The goal of this test file is to test the 2 public functions which are exported. Thus, I am mocking the API actions that these 2 functions use- I do this by having the API calls happen in private functions which can be mocked.

Reply:
I think you'd be able to accomplish this goal more if you followed this pattern with the getSDClient() function instead of the functions wrapping getSDClient(). 

There are several things (filtering, etc) that don't get covered in the tests when you mock at this level.

Reply:
Yeah, but then I have to create an entire mock object, and that's what I'm trying to avoid here. However, by popular demand I will change the `listNamespaces` function to not have any more logic than a super duper thin wrapper of the SDK call. I will refactor that logic out elsewhere so that it is tested. I think that should satisfy both of your concerns.

Reply:
I agree with the above comments that mocking the entire listNamespaces() as it currently appears hides too much logic (like filtering) from testing. Reducing the scope of this func to only encapsulate the actual API call (and testing the filter logic via tests) will go a long way in addressing my concerns.